modules/fancymeshrenderer/src/calcnormals.cpp

```diff
@@ -0,0 +1,152 @@
+                    return dvec3(v[i], 0, 0);
+                }
+                else if constexpr (util::extent<T>::value == 2) {
+                    return dvec3(v[i], 0);
+                } else {  // extent == 3 or extent == 4
+                    return dvec3(v[i]);
+                }
+            };
+        });
+
+    // loop over index buffers
+    for (auto [meshInfo, buffer] : mesh.getIndexBuffers()) {
+        if (meshInfo.dt != DrawType::Triangles) continue;
+        meshutil::forEachTriangle(meshInfo, *buffer, [&](auto i0, auto i1, auto i2) {
+            const auto v0 = vertexLookUp(i0);
+            const auto v1 = vertexLookUp(i1);
+            const auto v2 = vertexLookUp(i2);
+
+            const dvec3 n = cross(v1 - v0, v2 - v0);
+            double l = glm::length(n);
+            if (l < std::numeric_limits<float>::epsilon()) {
+                // degenerated triangle
+                return;
+            }
+            // weighting factor
+            double weightA;
+            double weightB;
+            double weightC;
+            switch (mode) {
+                case Mode::WeightArea:
+                    // area = norm of cross product
+                    weightA = 1;
+                    weightB = 1;
+                    weightC = 1;
+                    break;
+                case Mode::WeightAngle: {
+                    // based on the angle between the edges
+                    const dvec3 e0 = glm::normalize(v1 - v2);
+                    const dvec3 e1 = glm::normalize(v2 - v0);
+                    const dvec3 e2 = glm::normalize(v1 - v0);
+                    weightA = acos(dot(e1, e2)) / l;
+                    weightB = acos(dot(e0, e2)) / l;
+                    weightC = acos(dot(e0, e1)) / l;
+                    break;
+                }
+                case Mode::WeightNMax: {
+                    auto edge = [](auto a, auto b) {
+                        auto e = a - b;
+                        auto l = glm::length(e);
+                        return std::make_pair(e / l, l);
+                    };
+                    const auto [e0, l0] = edge(v1, v2);
+                    const auto [e1, l1] = edge(v2, v0);
+                    const auto [e2, l2] = edge(v1, v0);
+                    weightA = sin(acos(dot(e1, e2))) / (l * l1 * l2);
+                    weightB = sin(acos(dot(e0, e2))) / (l * l0 * l2);
+                    weightC = sin(acos(dot(e0, e1))) / (l * l0 * l1);
+                    break;
+                }
+                default:
+                    weightA = weightB = weightC = 1.0 / l;  // no weighting
+            }
+            // add it to the vertices
+            normals[i0] += vec3(n * weightA);
+            normals[i1] += vec3(n * weightB);
+            normals[i2] += vec3(n * weightC);
+        });
+    }
+
+    // normalize normals
+    std::transform(normals.begin(), normals.end(), normals.begin(), [](auto n) {
```

Comment:
should this be optional, with default to true?

Reply:
No, I don't see why it should. Since the new normals are a weighted average based on its neighboring faces we would then need to keep track of the total weight and divide by that instead, which is more costly.  And normals should always have the length 1